src/main/scala/is/hail/rvd/UnpartitionedRVD.scala

```diff
@@ -49,4 +49,26 @@ class UnpartitionedRVD(val rowType: TStruct, val rdd: RDD[RegionValue]) extends
   }
 
   def coalesce(maxPartitions: Int, shuffle: Boolean): UnpartitionedRVD = new UnpartitionedRVD(rowType, rdd.coalesce(maxPartitions, shuffle = shuffle))
+
+  def constrainToOrderedPartitioner(
+    ordType: OrderedRVDType,
+    newPartitioner: OrderedRVDPartitioner
+  ): OrderedRVD = {
+
+    assert(ordType.rowType == rowType)
+
+    val localRowType = rowType
+    val pkOrdering = ordType.pkType.ordering
+    val rangeTree = newPartitioner.rangeTree
+    val filtered = rdd.mapPartitions { it =>
+      val ur = new UnsafeRow(localRowType, null, 0)
+      val key = new KeyedRow(ur, ordType.pkRowFieldIdx)
+      it.filter { rv =>
+        ur.set(rv)
+        rangeTree.contains(pkOrdering, key)
+      }
+    }
+
+    OrderedRVD.shuffle(ordType, newPartitioner, filtered)
```

Comment:
we might want to consider using the coerce strategy with `newPartitioner` as a hint partitioner. Maybe not, though.

Reply:
`coerce` felt too high level for this, but I'm open to discussion. This method says "give me back an `OrderedRVD` with exactly this partitioner, dropping any data that fall outside the given partitions." So the partitioner argument is more than a hint.

The other factor is that `coerce` does a pass over the data collecting statistics, and detecting if it was already ordered. My feeling was that this method should be an explicit "just do a shuffle". Probably collecting statistics and making choices based on the results should be explicit in the IR, not built into the methods the IR calls. Maybe @cseed has thoughts?